mlir/lib/Dialect/Vector/Transforms/VectorTransforms.cpp

```diff
@@ -1622,7 +1622,33 @@ struct ChainedReduction final : OpRewritePattern<vector::ReductionOp> {
   }
 };
 
-/// For vectors with either leading or trailing unit dim, replaces:
+// Helper function dropping unit non-scalable dimension from a VectorType.
+// Scalable unit dimensions are not dropped. Folding such dimensions would
+// require "shifting" the scalable flag onto some other fixed-width dim (e.g.
+// vector<[1]x4xf32> -> vector<[4]xf32>). This could be implemented in the
+// future.
+static VectorType dropNonScalableUnitDimFromType(VectorType inVecTy) {
+  auto inVecShape = inVecTy.getShape();
+  SmallVector<int64_t> newShape;
+  SmallVector<bool> newScalableDims;
+  for (auto [dim, isScalable] :
+       llvm::zip_equal(inVecShape, inVecTy.getScalableDims())) {
+    if (dim == 1 && !isScalable)
+      continue;
+
+    newShape.push_back(dim);
+    newScalableDims.push_back(isScalable);
+  }
+  // All dims have been dropped, we need to return a legal shape for VectorType.
```

Comment:
nit: 0-D vectors are legal, but not well supported (I think that was the issue from IREE)?

Reply:
Yes. I saw some instances essentially with data transfer memory shape ops (extract, broadcast, ...). Do you think the solution rather allow elementwise op to support vector<f32> ?

It was indeed an issue coming from IREE yet, it was reproduced with LLVM

Reply:
We don't need to think about IREE here. We should generate valid IRs in upstream transformations. IMO, 0-D vector is weird. I can't really tell the distinction between 0-D vector and `vector<1xT>`. I'd suggest to return 1-D vector and document it in the function comment (i.e., l.1625-l.1629).

Reply:
+1 to what @hanhanW is suggesting, thanks!